finality-aleph/src/network/address_cache.rs

```diff
@@ -0,0 +1,180 @@
+use std::{collections::HashMap, marker::PhantomData, num::NonZeroUsize, sync::Arc};
+
+use lru::LruCache;
+use parking_lot::Mutex;
+use primitives::{AccountId, AlephSessionApi, AuthorityId, BlockHash, BlockNumber};
+use sc_client_api::Backend;
+use serde::{Deserialize, Serialize};
+use sp_runtime::traits::{Block, Header};
+
+use crate::{
+    abft::NodeIndex,
+    session::{SessionBoundaryInfo, SessionId},
+    session_map::AuthorityProvider,
+    ClientForAleph,
+};
+
+pub trait KeyOwnerInfoProvider {
+    fn aleph_key_owner(&self, block_number: BlockNumber, key: AuthorityId) -> Option<AccountId>;
+}
+
+/// Network details for a given validator. This data is purely informational
+/// and can change over time, even within a single session.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ValidatorAddressingInfo {
+    /// Network level address of the validator, i.e. IP address
+    pub network_level_address: String,
+    /// PeerId of the validator used in validator (clique) network
+    pub validator_network_peer_id: String,
```

Comment:
nit: change String to PeerId

Reply:
I'd rather not do it, moving discussion to #1449